source/vir/src/well_formed.rs

```diff
@@ -306,6 +318,102 @@ fn check_function(ctxt: &Ctxt, function: &Function) -> Result<(), VirErr> {
         }
     }
 
+    if function.x.attrs.integer_ring {
+        if function.x.mode != Mode::Proof {
+            return err_str(&function.span, "integer_ring mode must be declared as proof");
+        }
+        if let Some(body) = &function.x.body {
+            crate::ast_visitor::expr_visitor_check(body, &mut |expr| {
+                match &expr.x {
+                    ExprX::Block(_, _) => {}
+                    _ => {
+                        return err_str(&function.span, "integer_ring mode cannot have a body");
+                    }
+                }
+                Ok(())
+            })?;
+        }
+        for p in function.x.params.iter() {
+            match *p.x.typ {
+                TypX::Int(IntRange::Int) => {}
+                TypX::Boxed(_) => {}
+                _ => {
+                    return err_str(&p.span, "integer_ring mode's param should be int type");
+                }
+            }
+        }
+        if function.x.ensure.len() != 1 {
+            return err_str(
+                &function.span,
+                "integer_ring mode: single ensure expression is assumed now",
+            );
+        } else {
+            let ens = function.x.ensure[0].clone();
+            if let crate::ast::ExprX::Binary(crate::ast::BinaryOp::Eq(_), lhs, rhs) = &ens.x {
+                if let crate::ast::ExprX::Binary(
+                    crate::ast::BinaryOp::Arith(ArithOp::EuclideanMod, _),
+                    _,
+                    _,
+                ) = &lhs.x
+                {
+                    match &rhs.x {
+                        crate::ast::ExprX::Const(crate::ast::Constant::Nat(zero))
+                            if "0" == (**zero).as_str() => {}
+                        _ => {
+                            return err_str(
+                                &function.span,
+                                "integer_ring mode ensure expression: `X % m == 0` or `X == Y` ",
+                            );
+                        }
+                    }
+                }
+            } else {
+                return err_str(
+                    &function.span,
+                    "integer_ring mode ensures expression: operator `==` is assumed",
```

Comment:
What do you mean that's assumed? That it is required?

Reply:
I wanted to say that the most outer operator for `ensures` expression should be the equality operator  (not any functions that return boolean, especially shouldn't be inequalities)